source/normalize-arguments.js

```diff
@@ -108,14 +108,31 @@ module.exports = (url, options) => {
 		}
 	}
 
-	if (!is.function(options.retries)) {
-		const {retries} = options;
+	options.gotRetries = {retry: 0, methods: [], statusCodes: []};
+	if (is.function(options.retries) || is.number(options.retries)) {
+		options.gotRetries.retry = options.retries;
+	} else if (is.object(options.retries)) {
+		options.gotRetries = {...options.gotRetries, ...options.retries};
+	}
+	delete options.retries;
+
+	if (!is.function(options.gotRetries.retry)) {
+		const {retry} = options.gotRetries;
 
-		options.retries = (iter, error) => {
-			if (iter > retries || !isRetryAllowed(error)) {
+		options.gotRetries.retry = (iter, error) => {
+			if (iter > retry || (!isRetryAllowed(error) && !options.gotRetries.methods.includes(error.method) && !options.gotRetries.statusCodes.includes(error.statusCode))) {
 				return 0;
 			}
 
+			if (error.statusCode === 413 && Reflect.has(error.headers, 'retry-after')) {
+				const after = Number(error.headers['retry-after']);
```

Comment:
Keep in mind this comment: https://github.com/sindresorhus/got/issues/417#issuecomment-364760887 We need to set some sort of upper limit (and document it). What kind of limit do you think makes sense?

Reply:
1. There's limit of redirects set to 10.
2. We can set a timeout, if it exceeds the timeout then throw.

> We need to set some sort of upper limit (and document it). What kind of limit do you think makes sense?

Indeed. It should be the `retries` option.

Reply:
I was referring to a time limit, not redirect limit.

Reply:
> We can set a timeout, if it exceeds the timeout then throw.

Reply:
> We can set a timeout, if it exceeds the timeout then throw.

What should the timeout be? Should it be configurable?

How does retry interact with the set timeout?

Reply:
> What should the timeout be? Should it be configurable?

Yes, we should specify `retries: { ..., timeout: number }`. That would override the `timeout` option. By default it should be set to `15000` or `30000` I think.

> How does retry interact with the set timeout?

It just makes a new request and uses the timeout option provided by user.
If got 413 status code, then it overrides the timeout according to the `Retry-After` header (if exists). I think I should note that in the docs.

Reply:
> Yes, we should specify retries: { ..., timeout: number }. That would override the timeout option. By default it should be set to 15000 or 30000 I think.

So why wouldn't the user just change the `timeout` option themselves instead then? If `options.retries.timeout` just overrides `options.timeout`.

> I think I should note that in the docs.

Yes

Reply:
> So why wouldn't the user just change the timeout option themselves instead then? If options.retries.timeout just overrides options.timeout.

Reply:
> What should the timeout be? Should it be configurable?

It could be the already configured timeout. If `retry-after` > `options.gotTimeout.request`, then don't retry. Not sure I'd add a new timeout until it's needed.

Reply:
> Be aware Number(null) is 0, which would cause an immediate retry. Maybe not worth coding for, though.